provisioner/converge/provisioner.go

```diff
@@ -0,0 +1,240 @@
+// This package implements a provisioner for Packer that executes
+// Converge to provision a remote machine
+
+package converge
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"strings"
+
+	"encoding/json"
+
+	"regexp"
+
+	"github.com/mitchellh/packer/common"
+	"github.com/mitchellh/packer/helper/config"
+	"github.com/mitchellh/packer/packer"
+	"github.com/mitchellh/packer/template/interpolate"
+)
+
+var versionRegex = regexp.MustCompile(`^[\.\-\da-zA-Z]*$`)
+
+// Config for Converge provisioner
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+
+	// Bootstrapping
+	Bootstrap bool   `mapstructure:"bootstrap"`
+	Version   string `mapstructure:"version"`
+
+	// Modules
+	ModuleDirs []ModuleDir `mapstructure:"module_dirs"`
+
+	// Execution
+	Module           string            `mapstructure:"module"`
+	WorkingDirectory string            `mapstructure:"working_directory"`
+	Params           map[string]string `mapstucture:"params"`
+	ExecuteCommand   string            `mapstructure:"execute_command"`
+
+	ctx interpolate.Context
+}
+
+// ModuleDir is a directory to transfer to the remote system
+type ModuleDir struct {
+	Source      string   `mapstructure:"source"`
+	Destination string   `mapstructure:"destination"`
+	Exclude     []string `mapstructure:"exclude"`
+}
+
+// Provisioner for Converge
+type Provisioner struct {
+	config Config
+}
+
+// Prepare provisioner somehow. TODO: actual docs
+func (p *Provisioner) Prepare(raws ...interface{}) error {
+	err := config.Decode(
+		&p.config,
+		&config.DecodeOpts{
+			Interpolate:        true,
+			InterpolateContext: &p.config.ctx,
+			InterpolateFilter: &interpolate.RenderFilter{
+				Exclude: []string{"execute_command"},
+			},
+		},
+		raws...,
+	)
+	if err != nil {
+		return err
+	}
+
+	// require a single module
+	if p.config.Module == "" {
+		return errors.New("Converge requires a module to provision the system")
+	}
+
+	// set defaults
+	if p.config.WorkingDirectory == "" {
+		p.config.WorkingDirectory = "/tmp"
+	}
+
+	if p.config.ExecuteCommand == "" {
+		p.config.ExecuteCommand = "cd {{.WorkingDirectory}} && sudo converge apply --local --log-level=WARNING --paramsJSON '{{.ParamsJSON}}' {{.Module}}"
+	}
+
+	// validate version
+	if !versionRegex.Match([]byte(p.config.Version)) {
+		return fmt.Errorf("Invalid Converge version %q specified. Valid versions include only letters, numbers, dots, and dashes", p.config.Version)
+	}
```

Comment:
This is just me reflexively sanitizing input. Since it gets passed to a shell script, I was concerned that it might contain a quote or something that would break the system. Do you think it's unneeded here? The regex is quite broad (`^[A-Za-z\-\.]*$`).